core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/VariableReferenceGraph.java

```diff
@@ -0,0 +1,330 @@
+        return variableReferenceToInstanceMap.getOrDefault(variableId, Collections.emptyMap()).get(entity);
+    }
+
+    public void addFixedEdge(EntityVariablePair from, EntityVariablePair to) {
+        if (from.graphNodeId() == to.graphNodeId()) {
+            return;
+        }
+        fixedEdges.computeIfAbsent(from, k -> new ArrayList<>()).add(to);
+    }
+
+    public void addEdge(EntityVariablePair from, EntityVariablePair to) {
+        if (from.graphNodeId() == to.graphNodeId()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var oldCount = counts[from.graphNodeId()][to.graphNodeId()]++;
+        if (oldCount == 0) {
+            graph.addEdge(from.graphNodeId(), to.graphNodeId());
+        }
+
+        markChanged(to);
+    }
+
+    public void removeEdge(EntityVariablePair from, EntityVariablePair to) {
+        if (from.graphNodeId() == to.graphNodeId()) {
+            return;
+        }
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        var newCount = --counts[from.graphNodeId()][to.graphNodeId()];
+        if (newCount == 0) {
+            graph.removeEdge(from.graphNodeId(), to.graphNodeId());
+        }
+        markChanged(to);
+    }
+
+    public void markChanged(EntityVariablePair node) {
+        if (changed.isEmpty()) {
+            graph.startBatchChange();
+        }
+        changed.set(node.graphNodeId());
+    }
+
+    record AffectedEntity(Object entity, VariableUpdaterInfo variableUpdaterInfo) {
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof AffectedEntity other) {
+                return entity == other.entity;
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return System.identityHashCode(entity);
+        }
+    }
+
+    public void updateChanged() {
+        graph.endBatchChange();
+        var visited = new boolean[instanceList.size()];
+        var loopedTracker = new LoopedTracker(visited.length);
+        var affectedEntities = Collections.newSetFromMap(new IdentityHashMap<AffectedEntity, Boolean>());
+        var nodeHeap = createInitialChangeQueue();
+
+        while (!nodeHeap.isEmpty()) {
+            var nextNode = nodeHeap.poll().nodeId;
+            if (visited[nextNode]) {
+                continue;
+            }
+            visited[nextNode] = true;
+            var shadowVariable = instanceList.get(nextNode);
+            var isChanged = updateShadowVariable(shadowVariable,
+                    graph.isLooped(loopedTracker, nextNode),
+                    affectedEntities);
+
+            if (isChanged) {
+                graph.nodeForwardEdges(nextNode).forEachRemaining(
+                        (int node) -> nodeHeap.add(new AffectedShadowVariable(node, graph.getTopologicalOrder(node))));
```

Comment:
Performance idea as I was reading through this: could it be possible that a node is being added even though it is already `visited`? If so, let's not add it, save ourselves some iteration.